### ISSUE TITLE
prop-types: children and React.VFC

### ISSUE THREAD

**Issue by MOZGIII** [yannickcr/eslint-plugin-react#2913]
There is now a way to properly communicate the absence of the `children` prop at the TypeScript level: `React.VFC` (vs `React.FC`).
`React.VFC` aka `React.VoidFunctionComponent` does not allow `children` prop, the `prop-types` rule should derive the presence/absence of `children` from the underlying TypeScript types (when used with TypeScript).

Related: #7.

**Comment by ljharb** [comment]
That seems very wrong and strange to me; “void” refers to the return type generally, but “takes children” has no bearing on “renders children”. Can you confirm?

**Comment by MOZGIII** [comment]
Yes, the naming is poor, but the semantics is truly about taking children, and not about the "void".

**Comment by ljharb** [comment]
Who decided this type name? Can you provide a link?

**Comment by MOZGIII** [comment]
Here is the type: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f4f453b018d0d512cde50efb6db35fd3f8a001ab/types/react/index.d.ts#L551
Here's the PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46643

**Comment by ljharb** [comment]
Thanks.

While it is quite a terrible name, it makes sense that prop-types should be able to respect it.

**Comment by sladyn98** [comment]
@ljharb I could give this a shot. From what I understand is that the plugin needs to communicate the absence of children prop at the TypeScript level: `React.VFC (vs React.FC)`. Should the prop type rules be changed to accomodate this

**Comment by ljharb** [comment]
I'm not sure where the best place to change it is - we have shared propType detection code, and it might be useful to detect this there so other propType rules can check it. However, it may instead be better to only check it in the prop-types rule.

**Comment by sladyn98** [comment]
@ljharb I guess the `React.VFC` would be deprecated in the next release so adding it to the prop-type rules makes sense. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46643 We could get rid of the rule later ?

**Comment by ljharb** [comment]
What indicates it would be deprecated?

**Comment by sladyn98** [comment]
@ljharb Yeah your right there is no idication I guess I just read it somewehre

**Comment by tlmader** [comment]
@ljharb @sladyn98 ran into it here: https://basarat.gitbook.io/typescript/tsx/react#void-function-components

> As of @types/react PR #46643, you can use a new React.VoidFunctionComponent or React.VFC type if you wish to declare that a component does not take children. This is an interim solution until the next major version of the type defs (where VoidFunctionComponent will be deprecated and FunctionComponent will by default accept no children).

**Comment by ljharb** [comment]
In that case we’d need to know which version of the types we were looking at, which makes this a bit more complex :-/

**Comment by reberthkss** [comment]
Which class replaces `VFC`?

**Comment by flying-sheep** [comment]
I like that they’ll delete `FC` and rename `VFC`→`FC` in the next major release. That’s how it should have been from the start.

If that’s the intention, `VFC` is truly a horrible name though. it should be named `E`(xplicit)`FC` then: VFC implies the upgrade path

- use FC for components taking children, specify child type in props interface only if it’s not `ReactNode`
- use VFC for components not taking children
- sort out the mess when upgrading to the next major typing version

I much prefer this upgrade path:

- use `import { VFC as FC } from react` everywhere, explicitly state child types
- delete `VFC as ` in next major typing version
